### PR TITLE
Superagent: Enhance and update types to reflect superagent 8

### PR DESCRIPTION
**Pull request by DavidTanner**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  https://github.com/ladjs/superagent/blob/1c8338b2e0a3b8f604d08acc7f3cbe305be1e571/HISTORY.md
  https://github.com/ladjs/superagent/releases
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DavidTanner** [comment]
Frustrating...

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## supertest ([<kbd>unpkg</kbd>](https://unpkg.com/browse/supertest@latest/))
was missing the following properties:
1. Test

## Formatting

The following files are not formatted:
1. types/supertest-as-promised/index.d.ts
1. types/superagent/types.d.ts
1. types/superagent/index.d.ts
1. types/superagent/test.ts
1. types/supertest/supertest-tests.ts

<details>
<summary>as well as these 8 other files...</summary>
<p>types/supertest/types.d.ts, types/supertest/index.d.ts, types/superagent/lib/request-base.d.ts, types/supertest/lib/test.d.ts, types/supertest/lib/agent.d.ts, types/superagent/lib/node/agent.d.ts, types/superagent/lib/node/response.d.ts, types/superagent/lib/node/index.d.ts</p>
</details>


Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 671d4a2da0b5a8082035df4f76501f01e866d3a1
</p>

File: types/siren-parser/superagent.d.ts [hunk 1]
```diff
@@ -2,5 +2,5 @@ import { Response, SuperAgentRequest } from "superagent";
 import { Action, Entity } from "./index";
 
 export function parse(res: string): Entity;
-export function parse(res: Response, fn: (err: Error | null, body: Entity) => void): void;
+export function parse(res: Response, fn: (err: Error | null, body: any | Entity) => void): void;
```
Comment (dillonredding) [inline types/siren-parser/superagent.d.ts:5]: Union with `any` is like multiplying by 0.
```suggestion
export function parse(res: Response, fn: (err: Error | null, body: any) => void): void;
```

**Review by dillonredding — APPROVED** [review]

**Review by dillonredding — APPROVED** [review]

**Comment by andrewbranch** [comment]
@typescript-bot this has been approved by an owner

**Review by andrewbranch — APPROVED** [review]

**Comment by DavidTanner** [comment]
Ready to merge

**Comment by thynson** [comment]
The following code no longer be compiled any more:

```typescript
      const baseUrl = `http://localhost:${port}`;
      await supertest(baseUrl).get('/bar').expect(404);
```

[SuperTestStatic](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bccb4b09b56a0cbd4599ee99f158fe8767cf9e6b/types/supertest/index.d.ts#L32) does not accept string as its first param after this PR merged.

Probably can be fixed by

```diff
     interface SuperTestStatic {
         (app: App, options?: STAgentOptions): stAgent;
+        (url: string, options?: STAgentOptions): stAgent;
         Test: typeof STest;
         agent: typeof stAgent;
     }
```

**Comment by sjinks** [comment]
The following code does not compile anymore:

```js
request(server).get('/').set('X-Forwarded-For', '192.168.2.1').expect(403);
```

Error: Property 'set' does not exist on type 'Test'

**Comment by thynson** [comment]
> The following code does not compile anymore:
> 
> ```js
> request(server).get('/').set('X-Forwarded-For', '192.168.2.1').expect(403);
> ```
> 
> Error: Property 'set' does not exist on type 'Test'

Probably your case can be fixed by upgrading `@types/superagent` to latest version.

**Comment by sjinks** [comment]
@thynson: that worked, thanks! However, I did not have `@types/superagent` in my project; `@types/superagent` came from `@types/supertest`:

```
└─┬ @types/supertest@6.0.0
  └── @types/superagent@4.1.24
```

Looking at [`package.json`](https://github.com/DavidTanner/DefinitelyTyped/blob/97df53efac69ff69a78bb8819f52dfd75fda8486/types/supertest/package.json#L11), it feels like the version constraint for `@types/superagent` is not correct though.

**Comment by dehrhard** [comment]
> The following code no longer be compiled any more:


Facing the same issue

**Comment by ripecosta** [comment]
Changing the `@types/supertest`'s dependency on `@types/superagent` from `*` to `>= 8.0.0` would fix this automatically for everyone with needing an additional import in their package.json.

**Comment by DavidTanner** [comment]
I'll put up a PR for both these issues, unless someone beats me to it.

**Comment by strmer15** [comment]
This change also broke our type checking - looks like the `App` type is now requiring a `tls.Server` instead of an `http.Server`. The [ `supertest` docs say it should allow `http.Server`](https://github.com/ladjs/supertest?tab=readme-ov-file#example).

**Comment by viveleroi** [comment]
Am I missing something, why such a huge jump from v2 to v6? Updating these types breaks our typings, because `supertest.agent(url)` is no longer valid yet I see no mention of a breaking change in supertest releases, or even superagent (it's changelog is wanting). Has something changed in the api or are the types wrong?

`Value of type "typeof TestAgent` is not callable"

**Comment by DavidTanner** [comment]
@viveleroi can you post your code block, and I can get a PR up to fix it. 

The jump is to match the current version of supertest. These definitions haven't been updated in a very long time.

**Comment by stu43005** [comment]
I encountered an error while build:
```
node_modules/@types/superagent/lib/node/response.d.ts:13:26 - error TS2304: Cannot find name 'Request'.
13     constructor(request: Request);
                            ~~~~~~~
node_modules/@types/superagent/lib/node/response.d.ts:34:34 - error TS2304: Cannot find name 'Request'.
34     request: InstanceType<typeof Request>;
                                    ~~~~~~~
```

**Comment by viveleroi** [comment]
Here's our code block but most is building the URL. This file just exports the agent.

This works fine, but with v6 of the types throws an error `Value of type "typeof TestAgent" is not callable` but it works fine.

```tsx
import dotenv from 'dotenv'
import dotenvExpand from 'dotenv-expand'
import supertest from 'supertest'

dotenvExpand.expand(dotenv.config({ override: true }))

export const url = `${process.env.HOST}:${process.env.PORT}${process.env.CONTEXT_PATH}/api`

export const agent = supertest.agent(url)
```

**Pull request merged**